### PR TITLE
Add firstOf1, lastOf1, both1

### PR DESCRIPTION
**Pull request by phadej**
If this is ok, I'll add `maximumOf1`, `minimumOf1` etc `Semigroup` folds.

### CODE REVIEW

File: src/Control/Lens/Fold.hs [hunk 1]
```diff
@@ -599,6 +604,22 @@ toListOf :: Getting (Endo [a]) s a -> s -> [a]
...
+-- @
+-- 'toListOf' :: 'Getter' s a      -> s -> NonEmpty a
+-- 'toListOf' :: 'Fold1' s a       -> s -> NonEmpty a
+-- 'toListOf' :: 'Lens'' s a       -> s -> NonEmpty a
+-- 'toListOf' :: 'Iso'' s a        -> s -> NonEmpty a
+-- 'toListOf' :: 'Traversal1'' s a -> s -> NonEmpty a
+-- 'toListOf' :: 'Prism'' s a      -> s -> NonEmpty a
+-- @
+toNonEmptyOf :: Getting (NonEmpty a) s a -> s -> NonEmpty a
+toNonEmptyOf l = foldMapOf l (\x -> x :| [])
```
Comment (phadej) [inline src/Control/Lens/Fold.hs]: I'm not sure how to make dlist trick to avoid O(n^2) so went with this for now.

**Comment by phadej** [comment]
Side comment/question: would be cool to have `Apply` as superclass of `Applicative`, were there any related discussions somewhere?

**Comment by phadej** [comment]
travis breakage will be hopefully resolved via https://github.com/hvr/cassava/pull/125

**Comment by ekmett** [comment]
A large part of the reason for using the old cabals was to ensure we could still build with them without being caught by surprise. I'm happy to merge the first part.

**Comment by RyanGlScott** [comment]
Actually, it appears that `.travis.yml` is _currently_ using `cabal-install-head`, after 4b8751488fd1520a71f22ab910ca2542fb49dcf6.

**Comment by ekmett** [comment]
Then I guess we already took the plunge and will have to deal with these issues if we find them post-deployment.

**Pull request merged**